src/Microsoft.Identity.Web/TokenAcquisition.cs

```diff
@@ -435,20 +440,49 @@ public async Task RemoveAccountAsync(RedirectContext context)
         /// </summary>
         internal /* for testing */ async Task<IConfidentialClientApplication> GetOrBuildConfidentialClientApplicationAsync()
         {
+            // Use a double checking lock to avoid locking in most situations
             if (_application == null)
```

Comment:
> if (_application == null) [](start = 11, length = 26)

I would suggest to use the lazy load pattern instead

Reply:
Async lazy is a bit more complex, see https://blog.stephencleary.com/2012/08/asynchronous-lazy-initialization.html - but ultimately the code will be more readable.

Reply:
Async  lazy will not perform as well as this code, and this code is technically lazy loading. Only the first call takes the hit to load the _application. Subsequent calls will reuse that loaded value. The double checking lock is referring to the atomicity model rather than whether or not the data is lazy.

Reply:
Yup both will be lazy load, get the good old double locking. Trying to avoid the need to dispose. I don't find async lazy more complex. Both are fine, I just prefer the other one.

Reply:
I don't think Azync Lazy is needed, just Lazy as this is only for object initialization, not for the actual async operations. 

---
In reply to: [600648216](https://github.com/AzureAD/microsoft-identity-web/pull/1082#discussion_r600648216) [](ancestors = 600648216)

Reply:
I did some profiling. Anything under 4 parallel initializations will not benefit from lazy loading. Over 4 and a locked implementation starts to show benefit.